Bug 1431458 - add util that validates test main_summary output against prod

relud opened the pull request:
initially went with matching count and set of `document_id`s

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    // compare comparable columns
+    val originalColumns = original.selectExpr("document_id")
```
This can also just be `original.select`

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,94 @@
+    // compare row counts
+    val originalCount = original.count
+    val testCount = test.selectExpr("COUNT(*)").collect().last(0)
+    if (original.count != test.count) {
```
re-use the above variables

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,94 @@
+
+    // compare row counts
+    val originalCount = original.count
+    val testCount = test.selectExpr("COUNT(*)").collect().last(0)
```
Why not `test.count`?

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,94 @@
+      println("=======================================================================================")
+    }
+
+    spark.sparkContext.stop()
```
`spark.stop` also deals with all the relevant clean up.

relud commented:
apologies, i forgot to re-read the code after the tests were working, and I submitted the code in an incomplete state. I have corrected the review and added a check for nulls, and a relevant test.

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,93 @@
+
+  // Configuration for command line arguments
+  private class Conf(args: Array[String]) extends ScallopConf(args) {
+    val date = opt[String]("date", descr = "Submission date to compare", required = false)
```
This should be a required field

relud replied:
fixed

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,93 @@
+    val date = conf.date.get match {
+      case Some(f) => fmt.parseDateTime(f).toString(datePattern)
+      case _ => DateTime.now.minusDays(1).toString(datePattern)
+    }
```
This setup can be removed

relud replied:
fixed

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,93 @@
+
+    spark.stop()
+
+    if (fail) {
```
This variable could be replaced with the actual condition

relud replied:
fixed

relud commented:
@acmiyaguchi i added checks for null count in each column and mismatched columns.

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,97 @@
+    val original = spark.read.parquet(s"s3://$originalBucket/main_summary/$schemaVersion/submission_date_s3=$date")
+    val test = spark.read.parquet(s"s3://$testBucket/main_summary/$schemaVersion/submission_date_s3=$date")
+
+    if (!equal(date, original, test)) {
```
You could probably remove the date from this function, since all of the text will be wrapped in a box.

relud replied:
fixed

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,97 @@
+
+  // Configuration for command line arguments
+  private class Conf(args: Array[String]) extends ScallopConf(args) {
+    val date = opt[String]("date", descr = "Submission date to compare", required = true)
```
One last comment, this job would also be useful as a single sample_id or channel. I think passing in --channel=nightly is a common way of reducing the size of the main_summary job for validation.

relud replied:
added option `--where` and modified tests for improved coverage and ability to test `--where`

relud commented on src/test/scala/com/mozilla/telemetry/utils/MainSummaryValidateTest.scala:
```diff
@@ -0,0 +1,75 @@
+      "--date" :: date ::
+      "--original-viewname" :: original ::
+      "--test-viewname" :: test ::
+      {if (where == "") Nil else "--where" :: where :: Nil}
```
there is probably a more idiomatic way to express this, but i don't know it (yet)

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+    val original = conf.originalViewname.get match {
+      case Some(t) => spark.sql(s"SELECT * FROM $t").where(s"submission_date_s3 = '$date' and $where")
+      case _ => spark.read.parquet(s"s3://${conf.originalBucket()}/main_summary/$schemaVersion/submission_date_s3=$date").where(where)
+    }
```
I would actually do something like the following: 

```scala
val originalDF = where.get match {
  case Some(clause) => original.where(clause)
  case None => original
}
```

where original is the dataframe for main_summary. 

The argument is a Option[String], which is composed of two cases Some(x) and None. This means you don't need the default value of Some("TRUE") in the argument definition.

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+    same
+  }
+
+  def main(args: Array[String]) {
```
Can this be pushed up toward the top of the file?

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+    val originalBucket = opt[String]("original-bucket", descr = "Source bucket for original main_summary data", required = false, default = Some("telemetry-parquet"))
+    val originalViewname = opt[String]("original-viewname", descr = "View to pull origianl data from. Cannot be used with --original-bucket", required = false)
+    val testBucket = opt[String]("test-bucket", descr = "Source bucket for test main_summary data", required = false, default = Some("telemetry-test-bucket"))
+    val testViewname = opt[String]("test-viewname", descr = "View to pull test data from. Cannot be used with --test-bucket", required = false)
```
The test data is probably not going to be registered with hive, unfortunately.

relud replied:
right, this is basically just for unit tests, like GenericCountViewTest does.

fbertsch commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+  // Configuration for command line arguments
+  class Conf(args: Array[String]) extends ScallopConf(args) {
+    val date = opt[String]("date", descr = "Submission date to compare", required = true)
+    val originalBucket = opt[String]("original-bucket", descr = "Source bucket for original main_summary data", required = false, default = Some("telemetry-parquet"))
```
You can use `requiredOne` and `conflicts` to prevent the use of both this and `originalViewname`. See https://github.com/mozilla/telemetry-streaming/blob/master/src/main/scala/com/mozilla/telemetry/streaming/ErrorAggregator.scala#L83

fbertsch commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+
+    val original = conf.originalViewname.get match {
+      case Some(t) => spark.sql(s"SELECT * FROM $t").where(s"submission_date_s3 = '$date' and $where")
+      case _ => spark.read.parquet(s"s3://${conf.originalBucket()}/main_summary/$schemaVersion/submission_date_s3=$date").where(where)
```
This is main_summary specific, this could be a config

fbertsch commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+      case Some(t) => spark.sql(s"SELECT * FROM $t").where(s"submission_date_s3 = '$date' and $where")
+      case _ => spark.read.parquet(s"s3://${conf.originalBucket()}/main_summary/$schemaVersion/submission_date_s3=$date").where(where)
+    }
+    val test = conf.testViewname.get match {
```
This logic and the logic above (for `var original`) are the same, consider refactoring them together.

fbertsch commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    // check for extra columns
+    test.columns.foreach { col =>
```
Should this test for missing columns as well?

relud replied:
that happens a couple lines above during the null comparison check

fbertsch commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    // compare comparable columns
+    val originalColumns = original.select("document_id")
```
If these were configurable, it would generalize the ability to directly compare columns. It would then allow multiple, and let you know which ones were different and how different they were.

fbertsch commented on src/main/scala/com/mozilla/telemetry/utils/MainSummaryValidate.scala:
```diff
@@ -0,0 +1,103 @@
+    val spark = getOrCreateSparkSession(jobName)
+    val valid = validate(spark, conf)
+    spark.stop()
+    if (!valid) {
```
It might be good to have a config option of whether this will fail on failed validation. It might be a use case to have the job succeed, just report differences.

relud commented:
Addressed reviews

acmiyaguchi commented on src/main/scala/com/mozilla/telemetry/utils/DatasetComparator.scala:
```diff
@@ -0,0 +1,119 @@
+import org.rogach.scallop._
+
+object DatasetComparator {
+  def jobName: String = "main_summary_validate"
```
The name of the job has changed.

acmiyaguchi commented on src/test/scala/com/mozilla/telemetry/utils/DatasetComparatorTest.scala:
```diff
@@ -0,0 +1,77 @@
+class DatasetComparatorTest extends FlatSpec {
+  private val spark = SparkSession
+    .builder()
+    .appName("ClientCountViewTest")
```
Name should reflect the job.

fbertsch commented on src/main/scala/com/mozilla/telemetry/utils/DatasetComparator.scala:
```diff
@@ -0,0 +1,119 @@
+
+    val missing = originalColumns.except(testColumns).count
+    if (missing > 0) {
+      println(s"$missing missing row(s)")
```
Nit: change this verbiage, the rows aren't missing, just different.

relud replied:
I think that missing is the correct verbiage here. This shows distinct row values found in `original` and not `test`. The next check does distinct row values found in `test` and not `original`, and refers to them as extra rows. If only one of them is hit, then it definitely indicates missing or extra values.

That said, I think perhaps `values` would be more correct here, since it's a set operation, and only counts the number of distinct different values.
